Arc 23: Sharing Application information on the compiled application's bytes (#157)
algorandfoundation/ARCs: ARCs/arc-0023.md

> For the example above, the `bytecblock` is:
>
> ```
> bytecblock 0x6172633233017012202cf44ce287496f9080e1defd4474fa3b96ebbba8eefdd581acf8fa1a05e257b8
> ```
>
> and when compiled this gives the following byte string (at least with TEAL v8 and before):
>
> ```
> 0x2601296172633233017012202cf44ce287496f9080e1defd4474fa3b96ebbba8eefdd581acf8fa1a05e257b8
> ```

The size of the compiled application plus the bytecblock **MUST** be, at most, the maximum size of a compiled application according to the latest consensus parameters supported by the compiler.

> At least with TEAL v8 and before, appending the `bytecblock` to the end of the program should add exactly 44 bytes (1 byte for opcode `bytecblock`, 1 byte for 0x01 -the number of byte strings-, 1 byte for 0x29 the length of the encoded information byte string, 41 byte for the encodedin information byte string)

The `bytecblock` **MAY** be placed anywhere in the TEAL source code as long as it does not modify the semantic of the TEAL source code.

> However, if `approval.teal` is provided as an application information file, the `bytecblock` **SHOULD** be the last opcode of the TEAL program.

Reviewer: ```suggestion
However, if `approval.teal` is provided as an application information file, the `bytecblock` **SHOULD** be the last opcode of the deployed TEAL program.
```

---
arc: 23
title: Sharing Application's information
description: Append informations to the compiled application's bytes
author: Stéphane Barroso (@sudoweezy)
discussions-to: https://github.com/algorandfoundation/ARCs/issues/80
status: Draft
type: Meta
created: 2023-01-11
---

## Abstract

The following document introduces a convention for appending information to the compiled application's bytes. 
The goal of this convention is to standardize the process of verifying and interacting with smart contracts. The encoded byte is `arc23` followed by the length of the CID representing a folder of files & the CID itself.

## Specification

The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.

> Comments like this are non-normative.

### CID
The <a href="https://github.com/multiformats/cid">CID</a> allows to access the corresponding files using <a href="https://docs.ipfs.tech/">IPFS</a>.

> The CID **MUST** represent one file or a folder of files.

Reviewer 2: In addition, CID can have multiple format and hash algortihms.
I would recommend enforcing these.

```suggestion
The CID **MUST**:
* represent a folder of files
* be a version V1 CID
* use SHA-256 hash algorithm
```

> The minimum requirement for this ARC is a file representing the ABI (ABI as described in [ARC-4](arc-0004.md))

Reviewer 2: ```suggestion
The minimum requirement for this ARC is a folder containing a file `contract.json` representing the contract metadata (as described in the ABI in [ARC-4](arc-0004.md), and as extended by future potential ARCs).
```

Reviewer 2: The above comment is to add the ability to extend the contract metadata.
Also I think ABI is a general term defining methods, ...
while the specific metadata can be called "contract metadata" and named "contract.json".

Reviewer 2: This also follows convention there https://github.com/algorand-devrel/beaker/blob/master/examples/account_storage/DiskHungry.artifacts/contract.json

### Encoding
The encoded byte is `arc23` followed by the length of the CID, the number of files, and the CID itself.

eg: for the example stored in `/asset/arc-0023/Folder`, 
`arc23 59 2 bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici`
`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici` correspond to this Folder: `../assets/arc-0023/Folder`. It can be obtained with `ipfs add --cid-version=1 Folder/*  -q -w -n| tail -1`

If there is only one file, the CID **MUST** be related to the ABI of the application. 
eg: `arc23 59 1 bafkreiajguyi3i5hf4h7blslsv53rjbexzzvpkyhdf26hmuqthrk4iko4a`
`bafkreiajguyi3i5hf4h7blslsv53rjbexzzvpkyhdf26hmuqthrk4iko4a` correspond to this Abi File: `../assets/arc-0023/Folder/Abi.json`. It can be obtained with `ipfs add --cid-version=1 Folder/Abi.json -q`

> The option -n is for obtaining the hash without pinning the file, you **MAY** remove the option in production

Otherwise, if there is more than one file, the CID must correspond to a folder of files. The ABI **MUST** be accessible by querying and IPFS node with `{cid}/Abi.json`
eg:`bafybeihdzhfifv46p27ee6weqq6h7ydttzwni7gvdogr3mg6rwscdyzici/Abi.json`.

> ### Appending

Reviewer 2: ```suggestion
### Including the encoded information to the program
```

Reviewer 2: Also one question is whether the appending should be done at the end of the clear program and/or approval program.
Since clear programs are less space constrained in general, maybe we should put them there? What do you think?

> The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing 1-byte constant which is the encoded object.

Reviewer 2: ```suggestion
The encoded byte is appended to the compiled application as a <a href="https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/#bytecblock-bytes">bytecblock</a> containing `[]byte` constant which is the encoded object.
```

Reviewer 2: There are 3 properties we want of the way the information is included in the program:

1. Easy to extract: it should be easy to detect when ARC-23 information are added and extract them. In particular, this should not require the ability to fully parse TEAL code, nor find boundaries of op-codes. The number of false positive should be negligible and the code to parse should be easy enough.
2. Easy to securely add to programs: it should be easy to add to new programs in a fully automated and secure way. This means that adding it should not break security. In addition, the TEAL compiler should be able to handle this gracefully. See also https://github.com/algorandfoundation/ARCs/pull/80#issuecomment-1252662213
3. Short: it should not be too long.

If we enforce that the CID is always a V1 CID w/ SHA256, this makes it easier because then

```
bytecblock b64(YXJjMjMBcBIg48nKgteefr5CesSEPH/gc55s1HzVG40dsN6NpCHjKBI=)
```

is always encoded as (in hexadecimal):

`0x2601296172633233` followed by 36 bytes of the CID

(`0x26` is the `bytecblock` opcode, `0x01` is the number of []byte strings, `0x29` is the length = 41 bytes, `0x6172633233` is hexadecimal of `arc23)

If in addition it's always supposed to be at the end of the TEAL code, then it's very easy to parse:
go to 44 bytes before end of file.

Now, an alternative would be:

```
bytecblock "arc23" b64(CID)
```

is easier to parse in case the CID may have different length since then you can look for `0x2602056172633233` as prefix that is then followed by a varuint of the length of the CID and the CID.
This is easier than concatenating "arc23" to the CID and having a single []byte bytecblock element because we have a very easy long prefix (otherwise, it can get messy).

But if the length of the CID is fixed, then the first option saves 1 byte.

> The reason to use `bytecblock` is that adding a `bytecblock` opcode at the end of a TEAL program does not change the semantics of the program, as long as opcodes are correctly aligned, there is no jump after the last position (that would make the program fail without `bytecblock`), and there is enough space left to add the opcode.
The size of the compiled application + the bytecblock should be, at most, the maximum size of a compiled application according to the latest consensus parameters supported by the compiler.

> 

Reviewer 2: MAybe add a paragraph about non-deployed application.

### Use in development

In development, it may be preferable to not publish the application information on IPFS.
Tools that need these information (such as local block explorer) can request this information as a .zip or .tar.gz containing at the root all the required files.

## Rationale

> By appending the IPFS CID of the folder containing information about the Application, any user with access to the blockchain could easily verify the Application and the ABI of the Application and interact with it.

Reviewer 2: ```suggestion
By appending the IPFS CID of the folder containing information about the Application, any user with access to the blockchain could easily verify the Application and the contract metadata of the Application and interact with it.
```
